README.MD
# unocss-postcss-webpack-plugin


## unocss在vite中使用配置，关于unocss在vite中使用，自行查阅官网 https://unocss.dev/integrations/vite ,vite使用unocss就不会出现这样的问题

```js
//vite.config.ts
import { fileURLToPath, URL } from 'node:url'
import UnoCSS from 'unocss/vite'
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import vueJsx from '@vitejs/plugin-vue-jsx'


// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue(),
    vueJsx(),
    UnoCSS()
  ],
  resolve: {
    alias: {
      '@': fileURLToPath(new URL('./src', import.meta.url))
    }
  },
  css: {
    postcss: {
      plugins: [
          require('autoprefixer')({
              overrideBrowserslist: [
                'Android >= 6', 'iOS >= 10', 'ie >= 11', 'Firefox >= 35', 'chrome >= 40','safari >= 6'
              ]
          })
      ]
  },
  }
})

```
## 使用
```js
<template>
  <div>
    <div class="w-200px h-200px bg-gray-400 flex items-center justify-center">vite中使用unocss的写法</div>
    <div class="box">vite中没有使用unocss的写法</div>
  </div>
</template>

<script setup lang="ts">

</script>

<style scoped>
.box{
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: red;
  width: 200px;
  height: 200px;
}
</style>
```

## CSS展示
![1](https://raw.githubusercontent.com/UzumakiHan/static-files/master/images/unocss-postcss-webpack-plugin/vite-unocss-1.png)
![2](https://raw.githubusercontent.com/UzumakiHan/static-files/master/images/unocss-postcss-webpack-plugin/vite-unocss-2.png)


## unocss在webpack中使用配置，关于unocss在webpack中使用，自行查阅官网 https://unocss.dev/integrations/webpack
```js
const { defineConfig } = require('@vue/cli-service');
const UnoCSS = require('@unocss/webpack').default
module.exports = defineConfig({
    lintOnSave: false,
    transpileDependencies: true,
    configureWebpack: {
        plugins: [
            UnoCSS({})
        ],
    },
    css: {
        loaderOptions: {
            postcss: {
                postcssOptions: {
                  plugins: [
                    require('autoprefixer')({
                        overrideBrowserslist: [
                          'Android >= 6', 'iOS >= 10', 'ie >= 11', 'Firefox >= 35', 'chrome >= 40','safari >= 6'
                        ]
                    })
                  ],
                },
            },
        }
    },
});

```
```js
<template>
  <div>
    <div class="w-200px h-200px bg-gray-400 flex items-center justify-center">webpack使用unocss的写法</div>
    <div class="box">webpack没有使用unocss的写法</div>
  </div>
</template>

<script setup lang="ts">

</script>

<style scoped>
.box{
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: red;
  width: 200px;
  height: 200px;
}
</style>
```
## CSS展示
![1](https://raw.githubusercontent.com/UzumakiHan/static-files/master/images/unocss-postcss-webpack-plugin/webpack-unocss-1.png)
![2](https://raw.githubusercontent.com/UzumakiHan/static-files/master/images/unocss-postcss-webpack-plugin/webpack-unocss-2.png)

## 比较发现，在webpack中，unocss写法的css样式没有添加css前缀，详情可看这个issue  https://github.com/unocss/unocss/issues/386



## 使用unocss-postcss-webpack-plugin插件 webpack中，unocss 添加 postcss 编译 ,配合@unocss/webpack使用，前提已在项目中配置了unocss相关配置，以及项目中的css是单独打包成独立的css文件

## install

```
npm i unocss-postcss-webpack-plugin -D
```

## 使用
```js
//vue.config.js
const { defineConfig } = require('@vue/cli-service');
const UnoCSS = require('@unocss/webpack').default
const unocssPostcssWebpackPlugin=require('unocss-postcss-webpack-plugin')
module.exports = {
     configureWebpack: {
        plugins: [
            UnoCSS({}),
            unocssPostcssWebpackPlugin(), 
        ],
    },
    css:{
        loaderOptions: {
            postcss: {
                postcssOptions: {
                  plugins: [
                    require('autoprefixer')({
                        overrideBrowserslist: [
                          'Android >= 6', 'iOS >= 10', 'ie >= 11', 'Firefox >= 35', 'chrome >= 40','safari >= 6'
                        ]
                    })
                  ],
                },
            },
        },
        extract:true,//这个主要是设置单独打包css，
    }
}
```
## CSS展示
## unocss编写的css样式也加上了css前缀
![3](https://raw.githubusercontent.com/UzumakiHan/static-files/master/images/unocss-postcss-webpack-plugin/webpack-unocss-3.png)



### props

|   Prop    |   Type    |   Default  |   description    |   required    |
|  ----  | ----  |  ----  | ----  |  ----  | 
|   overrideBrowserslist |   Array<string>  |   ['Android >= 6', 'iOS >= 10', 'ie >= 1,'Firefox >= 35', 'chrome >= 40','safari >= 6']   |  browserslist    |   false   |



CHANGELOG.md
# Changelog

## [1.1.3](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/compare/v1.1.2...v1.1.3) (2023-10-16)


### 🐛 Bug Fixes | Bug 修复

* 优化打包压缩出现乱码 ([5771d29](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/5771d29cacbe541beeb485311356ddde7000ae13))


### 👷‍ Build System | 构建

* Update build dist ([4bd0f31](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/4bd0f31af3e9838326a03277ed4d43e6c2c0b6dc))

## [1.1.2](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/compare/v1.1.1...v1.1.2) (2023-10-07)


### 🐛 Bug Fixes | Bug 修复

* remove console ([5443e3e](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/5443e3e88b2e4c9e04bba988fd73471b8a5378a9))


### 👷‍ Build System | 构建

* Update build dist ([0ad5b97](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/0ad5b971aba49b2612c2a50f3c10e936bf7162b5))

## [1.1.1](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/compare/v1.1.0...v1.1.1) (2023-09-25)


### 🐛 Bug Fixes | Bug 修复

* 修复browserslist默认值 ([f2c542f](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/f2c542fa0034022564e7f2879e6b7fb0bbe6a67e))


### 👷‍ Build System | 构建

* Update build dist ([32db67f](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/32db67fa12295713b54994ac10b94add0bd77ed2))

## [1.1.0](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/compare/v1.0.0...v1.1.0) (2023-09-21)


### ✨ Features | 新功能

* update GITHUB_TOKEN ([e0e3cb6](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/e0e3cb63ac111f43bb4439748be161369b29ef95))

## 1.0.0 (2023-09-21)


### ✨ Features | 新功能

* 触发构建流程 ([18b4ffd](https://github.com/UzumakiHan/unocss-postcss-webpack-plugin/commit/18b4ffdc03ced6a2fa82e2e17bcd7601df08036e))


src/index.ts
import WebpackSources from 'webpack-sources';
import { createUnplugin } from 'unplugin';
import postcss from 'postcss';
import autoprefixer from 'autoprefixer';
const PLUGIN_NAME = 'unocss-postcss-webpack-plugin';
const DEFAULT_OVERRIDE_BROWSERS_LIST = [
    'Android >= 6', 'iOS >= 10', 'ie >= 11', 'Firefox >= 35', 'chrome >= 40','safari >= 6'
]
export function unocssPostcssWebpackPlugin(overrideBrowserslist: Array<string>=DEFAULT_OVERRIDE_BROWSERS_LIST) {
    return createUnplugin(() => {
        return {
            name: PLUGIN_NAME,
            enforce: 'post',
            webpack(compiler) {
                compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation: any) => {
                    compilation.hooks.optimizeAssets.tapPromise(PLUGIN_NAME, async () => {
                        const files = Object.keys(compilation.assets);
                        for (const file of files) {
                            if (!file.endsWith('.css')) {
                                continue;
                            }
                            const code = compilation.assets[file].source().toString();
                            try {
                                const result = await postcss([
                                    autoprefixer({
                                        overrideBrowserslist
                                    })
                                ]).process(code, {
                                    from: file
                                });
                                compilation.assets[file] = new WebpackSources.RawSource(result.content) as any;
                            } catch (e) {
                                console.error('[postcss autoprefixer error]:', e);
                            }
                        }
                    });
                });
            }
        };
    }).webpack();
}
